File: src/odin-libraries/python/odin.py
from os import environ
from sys import exit
from ruamel.yaml import YAML
import json
from odin_logger import OdinLogger as logger


class Odin:
    def __init__(self, config="job.yml", test=False):
        self.config = config
        try: 
            with open(self.config,"r") as config:
                configR = config.read()
            data = YAML().load(configR)
        except Exception as e:
            print(e)
        self.id = data["job"]["id"]

        if 'ODIN_EXEC_ENV' in environ or test != False:
            self.ENV_CONFIG = True
        else:
            self.ENV_CONFIG = False

    def condition(self, desc, expr):
        if self.ENV_CONFIG:
            logger.log("condition", desc, expr, self.id)
        return expr
    
    def watch(self, desc, value):
        if self.ENV_CONFIG:
            logger.log("watch", desc, value, self.id)

    def result(self, desc, status):
        if self.ENV_CONFIG:
            logger.log("result", desc, status, self.id)
        exit(0)

File: odin-libraries/nodejs/odin/odin.js
YAML = require('yamljs');
odinLogger = require('./odinLogger');

class Odin {
    constructor(config='job.yml', test=false) {
        this.config = YAML.load(config);
        this.test = test;
        this.id = this.config.job.id;
        // timestamp is used to identify approximate job run date time for observability
        this.timestamp = Date.now();
        this.logger = new odinLogger.OdinLogger();
        
        if (process.env.ODIN_EXEC_ENV) {
            this.ENV_CONFIG = true;
        }
        else {
            this.ENV_CONFIG = false;
        }
    }

    async condition(desc, expr){
        if (this.ENV_CONFIG) {
            this.logger.log('condition', desc, expr, this.id, this.timestamp);
        }
        return expr
    }

    async watch(desc, value){
         if (this.ENV_CONFIG) {
            this.logger.log('watch', desc, value, this.id, this.timestamp);
        }
    }

    async result(desc, status){
        if (this.ENV_CONFIG) {
            await this.logger.log('result',desc, status, this.id, this.timestamp);
        }
        process.exit(0);
    }
}

module.exports.Odin = Odin;